docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+    - **Data [optional]**
+        - Type: `*string`
+        - Description: Data contains the bootstrap data, such as cloud-init details scripts. If nil, the Machine should remain in the Pending state.
+
+## States and transitions
+The Cluster API Machine Controller is responsible for managing the lifecycle of a Machine, including its state transitions. This is the only controller that should be writing updates to Machines.
+
+---
+### Pending
+```go
+// MachinePhasePending is the first state a Machine is assigned by
+// Cluster API Machine controller after being created.
+MachinePhasePending = MachinePhaseType("pending")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is empty string
+
+#### Expectations
+- When `Machine.Spec.Bootstrap.Data` is:
+    - `<nil>`, expect the field to be set by an external controller.
+    - `""` (empty string), expect the bootstrap step to be ignored.
+    - `"..."` (populated by user or from the bootstrap provider), expect the contents to be used by a bootstrap (pre-Running transition) or infrastructure (Provisioning state) provider.
+
+---
+### Provisioning
+```go
+// MachinePhaseProvisioning is the state when the
+// Machine infrastructure is being created.
+MachinePhaseProvisioning = MachinePhaseType("provisioning")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Pending”
+- `Machine.Spec.Bootstrap.ConfigRef`->`Status.Phase` is “Ready”
+- `Machine.Spec.Bootstrap.Data` is not `<nil>`
+
+#### Expectations
+- Machine’s infrastructure to be in the process of being provisioned.
+
+---
+### Provisioned
+```go
+// MachinePhaseProvisioned is the state when its
+// infrastructure has been created and configured.
+MachinePhaseProvisioned = MachinePhaseType("provisioned")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Provisioning”
+- `Machine.Spec.InfrastructureRef`->`Status.Phase` is “Ready”
+
+#### Expectations
+- Machine’s infrastructure has been created and the compute resource is available to be configured.
+
+---
+### Running
+```go
+// MachinePhaseRunning is the Machine state when it has
+// become a Kubernetes Node in a Ready state.
+MachinePhaseRunning = MachinePhaseType("running")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Provisioned”
+- `Machine.Spec.ProviderID` is not `<nil>`
+- A Kubernetes Node is found with the same `Machine.Spec.ProviderID` and its state is “Ready”
+
+#### Expectations
+- Machine controller should set `Machine.Status.NodeRef`.
+
+---
+### Deleting
+```go
+// MachinePhaseDeleting is the Machine state when a delete
+// request has been sent to the API Server,
+// but its infrastructure has not yet been fully deleted.
+MachinePhaseDeleting = MachinePhaseType("deleting")
+```
+
+#### Transition Conditions
+- `Machine.Phase` can be any value
+- `Machine.ObjectMeta.DeletionTimestamp` is not `<nil>`
+
+#### Expectations
+- Node associated should be drained and cordoned.
```

Comment:
Implementation question: Will the machine controller do that or do we expect each provider implementation to do it?

Reply:
Ideally we should be doing this in a centralized manner, especially since the underlying "infrastructure" provider here would not really have knowledge of Kubernetes itself as it does today.

Reply:
This needs to be the responsibility of a singular cluster-api component, presumably the machine controller.

Reply:
OK. There's also a NodeMaintenance operator that the KubeVirt team might be willing to bring further upstream. https://github.com/kubevirt/node-maintenance-operator

Reply:
And as it happens, that is already in progress: https://github.com/kubernetes/enhancements/pull/1080

Reply:
As an end user, I want symmetry: If CAPI joins the node, I expect CAPI to gracefully remove that node, too.

At the KubeCon EU Contributors' Summit, I asked @lavalamp about the idea of a "[server-side drain](https://github.com/kubernetes/kubernetes/issues/25625)," and, from what I recall, he thought applying [taints](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#taint-based-evictions)  would have the desired effect.

I think we decided that the CAPI control plane has access to the workload cluster API. The Machine controller would have to (1) identify the node (it should be able to do that from the Machine object) and (2) apply Taint(s) to that node.